keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,604 @@
+
+Accordingly, new events related to this API should be added to [external-resizer](https://github.com/kubernetes-csi/external-resizer/blob/master/pkg/util/events.go):
+
+```
+VolumeModifyInProgress = "VolumeModifyInProgress"  // EventTypeNormal
+VolumeModifyFailed     = "VolumeModifyFailed"      // EventTypeWarning
+VolumeModifySuccess    = "VolumeModifySuccess"     // EventTypeNormal
+```
+
+5. Update quota code to use min and max(pvc.Spec.Resources, pvc.Status.AllocatedResources) for ProvisionedIOPS and ProvisionedThroughput when evaluating usage for PVC.
+
+![Modify Volume Flow](./qos_update_flow.png)
+
+Whenever resize controller or kubelet modifies pvc.Status (such as when setting both AllocatedResources and AllocatedResourceStatus) - it is expected that all changes to pvc.Status are submitted as part of same patch request to avoid race conditions.
+
+IO provisioning parameters can be considered as best-effort parameters, we should report the status of bad parameters and the volume would fall back to a workable default configuration.
+
+6. Add new operational and usage metrics to NodeGetVolumeStats:
+
+```go
+
+type Metrics struct {
+  // Existing
+  Time metav1.Time
+  Used *resource.Quantity
+  Capacity *resource.Quantity
+  Available *resource.Quantity
+
+  InodesUsed *resource.Quantity
+  Inodes *resource.Quantity
+  InodesFree *resource.Quantity
+
+  Abnormal *bool
+  Message *string
+
+  // For Provisioned IO KEP
+  IopsUsed *resource.Quantity
+  IopsAvailable *resource.Quantity
+
+  ThroughputUsed *resource.Quantity
+  ThroughputAvailable *resource.Quantity
+}
+
+```
+
+![QoS Flow Diagram](./qos_flow_diagram.png)
+
+### Test Plan
+
+* Basic unit tests for QoS and quota system.
+* E2e tests using mock driver to cause failure on create, update and recovering cases
+* Test coverage of quota usage with ResourceQuota and LimitRange
+
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+##### Integration tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature in 1.28 behind the VolumeProvisionedIO feature gate which is set to a default of false
```

Comment:
Will this be possible? I ask because this feature seems to depend on changes to the CSI spec

Reply:
This is open for discussion, I do not know what is the general timeline look like when we want to make change in CSI spec + API change(someone said it is possible but it is tight), the spec is in review https://github.com/container-storage-interface/spec/pull/544